src/lib.rs

```diff
@@ -0,0 +1,155 @@
+pub mod impls;
+
+pub trait ConstInit {
+    /// Create a new instance.
+    ///
+    /// This is a const instead of a method to allow creating instances in const context.
+    const INIT: Self;
+}
+
+/// Raw mutex trait.
+///
+/// This mutex is "raw", which means it does not actually contain the protected data, it
+/// just implements the mutex mechanism. For most uses you should use [`BlockingMutex`]
+/// instead, which is generic over a ScopedRawMutex and contains the protected data.
+///
+/// # Safety
+///
+/// ScopedRawMutex implementations must ensure that, while locked, no other thread can lock
+/// the RawMutex concurrently. This can usually be implemented using an [`AtomicBool`]
+/// to track the "taken" state. See [crate::impls] for examples of correct implementations.
+///
+/// Unsafe code is allowed to rely on this fact, so incorrect implementations will cause undefined behavior.
+///
+/// [`AtomicBool`]: core::sync::atomic::AtomicBool
+pub unsafe trait ScopedRawMutex {
+    /// Lock this `ScopedRawMutex`, calling `f()` after the lock has been acquired, and releasing
+    /// the lock after the completion of `f()`.
+    ///
+    /// If this was successful, `Some(R)` will be returned. If the mutex was already locked,
+    /// `None` will be returned
+    #[must_use]
+    fn try_lock<R>(&self, f: impl FnOnce() -> R) -> Option<R>;
+
+    /// Lock this `ScopedRawMutex`, calling `f()` after the lock has been acquired, and releasing
+    /// the lock after the completion of `f()`.
+    ///
+    /// Implementors may choose whether to block or panic if the lock is already locked.
+    /// It is recommended to panic if it is possible to know that deadlock has occurred.
+    ///
+    /// For implementations on a system with threads, blocking may be the correct choice.
+    ///
+    /// For implementations where a single thread is present, panicking immediately may be
+    /// the correct choice.
+    fn lock<R>(&self, f: impl FnOnce() -> R) -> R;
+}
+
+/// Blocking mutex (not async)
+///
+/// Provides a blocking mutual exclusion primitive backed by an implementation of [`ScopedRawMutex`].
+///
+/// Which implementation you select depends on the context in which you're using the mutex, and you can choose which kind
+/// of interior mutability fits your use case.
+///
+/// Use [`CriticalSectionRawMutex`] when data can be shared between threads and interrupts.
+///
+/// Use [`LocalRawMutex`] when data is only shared between tasks running on the same executor.
+///
+/// Use [`ThreadModeRawMutex`] when data is shared between tasks running on the same executor but you want a global singleton.
+///
+/// In all cases, the blocking mutex is intended to be short lived and not held across await points.
+///
+/// [`CriticalSectionRawMutex`]: crate::impls::cs::CriticalSectionRawMutex
+/// [`LocalRawMutex`]: crate::impls::local::LocalRawMutex
+/// [`ThreadModeRawMutex`]: crate::impls::thread_mode::ThreadModeRawMutex
+pub struct BlockingMutex<R, T: ?Sized> {
```

Comment:
Should maybe have a `Debug` impl when `T` and `R` are `Debug`? it could be feature-flagged for MCU targets that want to avoid unused debug impls...

We can always add that later.

Reply:
I think this would be good for a follow-on PR